### ISSUE TITLE
Allow having instance methods of the same name on different types in FFI

### ISSUE THREAD

**Issue by 2xsaiko** [dtolnay/cxx#471]
This currently doesn't work and fails with an "error[E0428]: the name \`__by_index\` is defined multiple times" even though both the C++ and method implementation code have non-ambiguous names:

```rust
#[cxx::bridge(namespace = "mcrtlib::ffi")]
mod types {
    pub struct TypeA {
        // (...)
    }

    pub struct TypeB {
        // (...)
    }

    extern "Rust" {
        fn by_index(self: &TypeA);
        fn by_index(self: &TypeB);
    }
}
```

**Comment by david-cattermole** [comment]
Hi @dtolnay,

I have a fix for this issue with my commit here: https://github.com/dtolnay/cxx/commit/a64bfd92fd7abee9b6c40a10ba5dcd19974761d6

My fix is to prefix the generated Rust function name with the struct name. I end up with the below expanded function names for two structs `ReadOperation` and `WriteOperation` that have the same method name `get_id`.

My test functions in my [example project branch](https://github.com/david-cattermole/cxx-demo-example/tree/instance-types-and-methods-with-same-name) now work using this new commit.

# Before
```rust
#[doc(hidden)]
#[export_name = "mmscenegraph$cxxbridge1$ReadOperation$get_id"]
unsafe extern "C" fn __get_id(__self: &ReadOperation) -> u8 {
    let __fn = "mmscenegraph_rs::cxxbridge::ffi::get_id";
    fn __get_id(__self: &ReadOperation) -> u8 {
        ReadOperation::get_id(__self)
    }
    ::cxx::private::catch_unwind(__fn, move || __get_id(__self))
}
#[doc(hidden)]
#[export_name = "mmscenegraph$cxxbridge1$WriteOperation$get_id"]
unsafe extern "C" fn __get_id(__self: &WriteOperation) -> u8 {
    let __fn = "mmscenegraph_rs::cxxbridge::ffi::get_id";
    fn __get_id(__self: &WriteOperation) -> u8 {
        WriteOperation::get_id(__self)
    }
    ::cxx::private::catch_unwind(__fn, move || __get_id(__self))
}
```

# After
```rust
#[doc(hidden)]
#[export_name = "mmscenegraph$cxxbridge1$ReadOperation$get_id"]
unsafe extern "C" fn __ReadOperation_get_id(__self: &ReadOperation) -> u8 {
    let __fn = "mmscenegraph_rs::cxxbridge::ffi::get_id";
    fn __ReadOperation_get_id(__self: &ReadOperation) -> u8 {
        ReadOperation::get_id(__self)
    }
    ::cxx::private::catch_unwind(__fn, move || __ReadOperation_get_id(__self))
}
#[doc(hidden)]
#[export_name = "mmscenegraph$cxxbridge1$WriteOperation$get_id"]
unsafe extern "C" fn __WriteOperation_get_id(__self: &WriteOperation) -> u8 {
    let __fn = "mmscenegraph_rs::cxxbridge::ffi::get_id";
    fn __WriteOperation_get_id(__self: &WriteOperation) -> u8 {
        WriteOperation::get_id(__self)
    }
    ::cxx::private::catch_unwind(__fn, move || __WriteOperation_get_id(__self))
}
```

**Comment by dtolnay** [comment]
I've released cxx-build 1.0.7 with a fix.
